src/matchbox/match_index.py

```diff
@@ -0,0 +1,218 @@
+    | Ob5    | Including | 1, 7     |
+    +--------+-----------+----------+
+
+    Should result in this index:
+
+    +-----------+---------------------+
+    | Trait     | Mismatched entities |
+    +===========+=====================+
+    | 1         | Ob1, Ob2            |
+    +-----------+---------------------+
+    | 2         | Ob1, Ob2, Ob5       |
+    +-----------+---------------------+
+    | 3         | Ob5                 |
+    +-----------+---------------------+
+    | 4         | Ob5                 |
+    +-----------+---------------------+
+    | 5         | Ob2, Ob3, Ob5       |
+    +-----------+---------------------+
+    | 6         | Ob2, Ob3, Ob5       |
+    +-----------+---------------------+
+    | 7         |                     |
+    +-----------+---------------------+
+    | Any other | Ob2, Ob5            |
+    +-----------+---------------------+
+    """
+
+    def __init__(self):
+        """Initialize the index."""
+        self.mismatch_unknown = set()
+        """
+        This set will keep included matching entities. They do not match unknown values.
+
+        Used for `self.index` default value, that means any previously unknown index value.
+        """
+        self.index = defaultdict(self.mismatch_unknown.copy)
+
+    def add_mismatch(self, indexed_entity, *traits_indexed_by):
+        """
+        Add a mismatching entity to the index.
+
+        We do this by simply adding the mismatch to the index.
+
+        :param collections.Hashable indexed_entity: an object to be mismatching the values of `traits_indexed_by`
+        :param list traits_indexed_by: a list of hashable values to index the object with
+        """
+        # If object is not matching given characteristic values, we should add it directly to the index.
+        for trait in traits_indexed_by:
+            self.index[trait].add(indexed_entity)
+
+    def add_match(self, indexed_entity, *traits_indexed_by):
+        """
+        Add a matching entity to the index.
+
+        We have to maintain the constraints of the data layout:
+            - `self.mismatch_unknown` must still contain all matched entities
+            - each key of the index must mismatch all known matching entities except those this particular key
+              explicitly includes
+
+        For data layout description, see the class-level docstring.
+
+        :param collections.Hashable indexed_entity: an object to be matching the values of `traits_indexed_by`
+        :param list traits_indexed_by: a list of hashable values to index the object with
+        """
+        # The index traits of `traits_indexed_by` might have already been used to index some other rntities. Those
+        # relations are to be preserved. If the trait was not used to index any entity, we initialize them to mismatch
+        # all matching entities known so far.
+        for trait in traits_indexed_by:
+            self.index.setdefault(trait, self.mismatch_unknown.copy())
```

Comment:
setdefault made it dramatically slow to create boxes (check out the performance pull-request, and compare total run times with setdefault, and relying on defaultdict to set up default values instead.

If you don't want to rely on defaultdict behaviour, the `if trait not in self.index:` condition would be much better. Otherwise you are copying mismatch at least as many times as matching entities are being added. and more if characteristic has many more traits.